<img src="https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Módulo** | Análise de Dados: Fundamentos de Estatística
Caderno de **Exercícios**<br> 
Professor [André Perez](https://www.linkedin.com/in/andremarcosperez/)

---

# **Tópicos**

<ol type="1">
  <li>Média e Variância;</li>
  <li>Ordem e Posição;</li>
  <li>Correlação.</li>
</ol>

---

# **Exercícios**

## 1\. Tráfego de São Paulo

Neste exercício, vamos continuar a analisar os dados de mobilidade urbana da cidade de São Paulo. A base de dados está neste [link](https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv) e é uma copia do dado original, presente neste [link](https://archive.ics.uci.edu/ml/datasets/Behavior+of+the+urban+traffic+of+the+city+of+Sao+Paulo+in+Brazil). A base de dados contem a quantidade de acidentes ocorridos na cidade entre 14/12/09 e 18/12/09, das 07:00h ás 20:00h, agregados em intervalos de 30 minutos.

In [ ]:
!wget -q "https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/develop/dataset/traffic.csv" -O traffic.csv

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('dados/traffic.csv', sep=';')

In [3]:
df.head()

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
0     1           0             0               0                0   
1     2           0             0               0                0   
2     3           0             0               0                0   
3     4           0             0               0                0   
4     5           0             0               0                0   

   running_over  fire_vehicles  occurrence_freight  \
0             0              0                   0   
1             0              0                   0   
2             0              0                   0   
3             0              0                   0   
4             0              0                   0   

   incident__dangerous_freight  lack_electricity  fire  flooding  \
0                            0                 0     0         0   
1                            0                 0     0         0   
2                            0                 0     0         0   
3                            0                 0     0         0   
4                            0                 0     0         0   

   manifestations  defect_trolleybuses  tree_road  semaphore_off  \
0               0                    0          0              0   
1               0                    0          0              0   
2               0                    0          0              0   
3               0                    0          0              0   
4               0                    0          0              0   

   intermittent_emaphore slowness_traffic_%  
0                      0                4,1  
1                      0                6,6  
2                      0                8,7  
3                      0                9,2  
4                      0               11,1

- **Exemplo**: (para relembrar)

O código abaixo extrai a 13ª linha do arquivo que representa a 14ª meia hora do dia 14/12/09, contadas a partir das 07:00h, ou seja, todos os incidentes ocorridos na cidade no dia em questão entre as 13:30h e 14:00h. Vemos, por exemplo, que dois ônibus e dois caminhões ficaram paradados nas ruas da cidade, atrapalhando o tráfego.

In [4]:
df.iloc[[13]]

hour  bus_stoped  broken_truck  vehicle_excess  accident_victim  \
13    14           2             2               0                1   

    running_over  fire_vehicles  occurrence_freight  \
13             0              0                   0   

    incident__dangerous_freight  lack_electricity  fire  flooding  \
13                            0                 0     0         0   

    manifestations  defect_trolleybuses  tree_road  semaphore_off  \
13               0                    0          0              0   

    intermittent_emaphore slowness_traffic_%  
13                      0                4,9

### **1.1. Agregação** 

Neste primeira atividade, você deve gerar um array NumPy por dia. Para cada array você deve somar todos os incidentes que aconteceram naquela meia hora. Sendo assim, cada array deve ter 27 posições, cada qual com a soma dos incidentes daquela meira hora.

**Dica**: Você deve remover a primeira e a última coluna.

**Dica**: Os arrays devem ter os seguintes valores:

```python
14: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 5, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 2, 3]
15: [3, 0, 0, 0, 2, 4, 2, 1, 3, 0, 2, 0, 2, 3, 2, 2, 4, 7, 6, 3, 0, 1, 1, 5, 4, 6, 0]
16: [1, 1, 2, 4, 4, 5, 3, 2, 4, 2, 1, 1, 3, 6, 2, 0, 2, 8, 4, 6, 4, 1, 2, 7, 15, 4, 0]
17: [4, 1, 1, 0, 0, 0, 2, 3, 2, 7, 2, 1, 7, 3, 6, 4, 3, 2, 2, 3, 2, 5, 2, 0, 1, 4, 15]
18: [10, 1, 1, 5, 5, 7, 3, 4, 3, 1, 2, 3, 2, 2, 6, 2, 2, 4, 2, 3, 0, 3, 3, 2, 3, 6, 0]
```

In [141]:
import numpy as np

df2 = df.iloc[:,1: -1]
x = np.array(df2)

day = 14
half = 0
incident_by_day = dict()
soma = list()

for y in x:
    try:
        soma.append(np.sum(y))
        half = half + 1
        if half == 27:
            incident_by_day[day] = soma
            day = day + 1
            half = 0
            soma = []
    except ValueError:
        continue

for day in incident_by_day:

  print(f'{day}: {incident_by_day[day]}')

14: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 5, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 2, 3]
15: [3, 0, 0, 0, 2, 4, 2, 1, 3, 0, 2, 0, 2, 3, 2, 2, 4, 7, 6, 3, 0, 1, 1, 5, 4, 6, 0]
16: [1, 1, 2, 4, 4, 5, 3, 2, 4, 2, 1, 1, 3, 6, 2, 0, 2, 8, 4, 6, 4, 1, 2, 7, 15, 4, 0]
17: [4, 1, 1, 0, 0, 0, 2, 3, 2, 7, 2, 1, 7, 3, 6, 4, 3, 2, 2, 3, 2, 5, 2, 0, 1, 4, 15]
18: [10, 1, 1, 5, 5, 7, 3, 4, 3, 1, 2, 3, 2, 2, 6, 2, 2, 4, 2, 3, 0, 3, 3, 2, 3, 6, 0]


### **1.2. Métricas** 

Para cada array você deve calcular as seguintes métricas: 

 - **média**;
 - **desvio padrão**.

In [170]:
for day in incident_by_day:
    print(f"O dia {day} tem a média de {np.mean(incident_by_day[day]):.2f} e o desvio de {np.std(incident_by_day[day]):.2f}.")

O dia 14 tem a média de 0.59 e o desvio de 1.19.
O dia 15 tem a média de 2.33 e o desvio de 2.02.
O dia 16 tem a média de 3.48 e o desvio de 3.05.
O dia 17 tem a média de 3.04 e o desvio de 3.05.
O dia 18 tem a média de 3.15 e o desvio de 2.19.


### **1.3. Interpretação** 

Baseado nos resultados da questão 1.2, responda:

> Qual dia apresenta a maior média de acidentes por meia hora?

**Resposta:** O dia 16 com uma média de 3.48 incidentes.

> Qual dia apresenta a menor variação de acidentes por meia hora?

**Resposta:** O dia 14 com um desvio padrão de 1.19

---